Add support for both plain and regex file names in create_permission.py

pthierry38 opened the pull request:
The goal of this patch is to support regex-based filename (only filename, not filepath), in the same csv file used to generate all file permission check oval files.

When the filename is a regex, the string must start with '^' and finish with '$' (matching a complete string). e.g. :
/boot,^System\\.map\.\*$,0,0,0600
to match any /boot/System.map* file.

I've tried to keep the oval xml filename and oval id as clean as possible. Yet the oval comment keeps the regex syntax.

The script has been test on the Ubuntu1604 csv file for this very file.

PS: is tere a doc file somewhere to specify all these files syntax for other users ?

pthierry38 commented:
One thing: this path does not support the ',' special char, due to the csv usage. It may be supported if the csv parsing takes care of the ^$ as delimiters, allowing syntaxes such as {2,} in the regex.

redhatrises commented on Ubuntu/16.04/input/profiles/anssi_np_nt28_average.xml:
```diff
@@ -35,7 +35,7 @@
 
 <!-- critical files -->
-<select idref="file_permissions_systemmap" selected="true"/>
+<select idref="file_permissions_boot_System_map" selected="true"/>
```
@pthierry38 please make the `S` in system lowercase.

redhatrises commented on shared/templates/template_BASH_regex_permissions:
```diff
@@ -0,0 +1,6 @@
+# strategy = configure
+# complexity = low
+# disruption = low
+find FILEPATH -regex '^FILEPATH/FILENAME' -exec chmod FILEMODE {} \;
```
Why not just extend the existing `template_BASH_permissions`? And why is `chmod 0600 /boot/System.map.*` not sufficient? I don't think filepaths are really going to need crazy regex that chmod cannot handle.

pthierry38 replied:
because the filepath is regex based, not globbing based. chmod is globbing based. If I just update the chmod, the regex will not work out of the box

pthierry38 replied:
As far as I understand, the unix:filename operation="pattern match" is regex based, but maybe it is possible to use only globbing :-) In that case I can update the mechanism to use globbing only... and to simplify the chmod directly...
What is harder with globbing is to differentiate in the csv a globbing name (foo*) with a real file named foo\* which is not a globbing. Using ^ and $ is easier to detect regex (but regex specific). But maybe I consider complicated useless cases ... ;-)

redhatrises replied:
I am still not seeing why regex here is needed where `chmod 0600 /boot/System.map.*` makes more sense. Now if you are wanting to find all files that have a specific or unique extension recursively across different file paths and change permissions, then I get using the `find` command with regex.

> In that case I can update the mechanism to use globbing only... and to simplify the chmod directly...

Please do that. This seems a little over-engineered for something that a simple chmod can do.

pthierry38 replied:
I've checked the oval definition. The operation="pattern match" supports only regex syntax, which means I can't write file names using globbing without translate the syntax from globbing to regex before generating the oval file. It looks complicated...

pthierry38 commented:
Upercase in oval id corrected. Also corrected for non-regex based.

mpreisler commented:
@redhatrises Any further feedback here? This is great work and I'd hate to have it rot.

pthierry38 commented:
Hi @mpreisler, @redhatrises,

To explain the need associated to this MR: there is some files for which the access rights must be checked according to the ANSSI NT-28 security policy which are suffixed under Debian with the current kernel version (system.map-kernvers-pkgvers). Depending on the install, there can be one ore multiple of this files in the same dir. That's why I've updated the syntax.

There is other cases of files suffixed by current release versions which makes arder to name them in oval file without globbing/regex. It's sad that OVAL pattern_match doesn't support globbing, but without any globbing/regex, all these oval files will need to be written "by hand" and updated at each new package update (due to the pkgvers value in the filename)

redhatrises commented:
That makes way more sense. Ack.
